### PR TITLE
[do-not-review] [cudagraph] Reuse capture-owned parameter gradients

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* (to be filled)

## Why

A full forward-backward CUDA graph allocates gradient storage during capture. For the single-backward path, retaining separate persistent gradients allocated during eager warmup duplicates that storage for the rest of training.

Gradient tensors allocated by capture must remain reachable after replay so eager gradient clipping and the optimizer can consume them, even though replay does not rerun the Python assignments to `parameter.grad`.

## Gradient ownership

```text
Before

Eager warmup gradients ------------------------------+
                                                      +--> persistent storage
CUDA graph capture gradients in the graph pool ------+

After

zero_grad(set_to_none=True)
           |
           v
CUDA graph capture allocates parameter.grad in its pool
           |
           v
record parameter -> captured gradient references
           |
           +--> graph replay writes gradient values
           +--> restore parameter.grad references
           +--> eager clipping and optimizer consume gradients
           +--> zero_grad(set_to_none=True) before the next replay
```

When gradient accumulation requires more than one backward call per optimizer step, TorchTitan preserves the existing persistent accumulation buffers instead.

## Change

Add `_GraphOwnedGradients` to record gradients allocated during capture and restore their parameter bindings after every replay. Validate that gradients are cleared before capture and replay and reject an unexpected gradient identity change.

Select `zero_grad(set_to_none=True)` only when CUDA graphs are disabled or the single-backward graph-owned-gradient path is active. Release the retained references during CUDA graph teardown.

## Testing

The tests cover capture-time recording, replay restoration, required gradient clearing, teardown, duplicate parameters, and Trainer policy for eager, graph, and gradient-accumulation modes.

- `python -m pytest -q tests/unit_tests/cpu/test_cudagraph.py tests/unit_tests/cpu/test_trainer.py`
- Four-GPU `pp_looped_1f1b_cudagraph` integration through capture and replay, 10/10 steps
- Deterministic four-GPU eager-versus-graph comparison with seed 42: bitwise-identical loss and `grad_norm` for all 10 steps

stack-info: PR: https://github.com/pytorch/torchtitan/pull/4436, branch: xmfan/stack/53

### CODE REVIEW

**Comment by xmfan** [comment]
Closing duplicate PR created by an incorrect ghstack resubmission; the original stack remains in #4429-#4436.
